mlflow/sagemaker/__init__.py

```diff
@@ -316,17 +371,75 @@ def _upload_s3(local_model_path, bucket, prefix):
+def _deploy(role, image_url, app_name, model_s3_path, run_id, region_name, mode, archive):
     """
     Deploy model on sagemaker.
     :param role: SageMaker execution ARN role
     :param image_url: URL of the ECR-hosted docker image the model is being deployed into
     :param app_name: Name of the deployed app
     :param model_s3_path: s3 path where we stored the model artifacts
     :param run_id: RunId that generated this model
+    :param mode: The mode in which to deploy the application.
+    :param archive: If True, any pre-existing SageMaker application resources that become inactive
+                    (i.e. as a result of deploying in `replace` mode) will be preserved. If False,
+                    these resources will be deleted.
     """
     sage_client = boto3.client('sagemaker', region_name=region_name)
-    model_name = app_name + '-model'
+    s3_client = boto3.client('s3', region_name=region_name)
+
+    deployed_endpoints = sage_client.list_endpoints()["Endpoints"]
+    deployed_endpoints = [endp["EndpointName"] for endp in deployed_endpoints]
+
+    endpoint_exists = (app_name in deployed_endpoints)
+    if endpoint_exists and mode == DEPLOYMENT_MODE_CREATE:
+        raise Exception("You are attempting to deploy application with name: {an} in `create` mode. However, an application with the same name"
+                        " already exists. If you want to update this application, deploy in `{madd}` or `{mrep}` mode".format(an=app_name,
+                                                                                                                              madd=DEPLOYMENT_MODE_ADD,
+                                                                                                                              mrep=DEPLOYMENT_MODE_REPLACE))
+    
+    elif endpoint_exists:
+        return _update_sagemaker_endpoint(endpoint_name=app_name,
+                                          image_url=image_url,
+                                          model_s3_path=model_s3_path,
+                                          run_id=run_id,
+                                          mode=mode, 
+                                          archive=archive,
+                                          role=role,
+                                          sage_client=sage_client,
+                                          s3_client=s3_client)
+
+    else:
+        return _create_sagemaker_endpoint(endpoint_name=app_name,
+                                          image_url=image_url,
+                                          model_s3_path=model_s3_path,
+                                          run_id=run_id,
+                                          role=role,
+                                          sage_client=sage_client,
+                                          s3_client=s3_client)
+
+
+def _get_sagemaker_model_name(endpoint_name):
+    dt = datetime.now()
+    return "{en}-model-{ts}".format(en=endpoint_name,
```

Comment:
Maybe we can use a UUID instead of a timestamp here to reduce the (probably pretty unlikely) odds of a collision e.g. if two users are simultaneously adding models to an endpoint? If we do want to use timestamps, we should increase the granularity of the timestamp (looks like these timestamps are in seconds, so a collision is maybe not totally improbable)